### PR TITLE
Move some AD calls into shared code

### PR DESCRIPTION
**Pull request by bill-long**
* Add Get-ActiveDirectoryAcl for retrieving AD ACLs without
  depending on the ActiveDirectory module and PSDrive.
* Add Get-ExchangeContainer, Get-OrganizationContainer, and
  Search-AllActiveDirectoryDomains to remove repetitive code
  from various scripts.

### CODE REVIEW

**Comment by dpaulson45** [comment]
We do a lot of `System.DirectoryServices.DirectorySearcher` maybe have some more options to handle those scenarios as well.

**Comment by bill-long** [comment]
I looked at DirectorySearcher, but all the uses are so different that I end up with something like Get-DirectorySearcher -Filter _something_ -PropertiesToLoad _something_ -CacheResult _something_. It doesn't really simplify things much. Open to ideas if you find a good way to do it, but that can be a future PR.

**Comment by lusassl-msft** [comment]
Validated the AD work in lab and it looks pretty good. Updated the `groupLists` one more time. 

Before /PrepareAllDomains
image

After /PrepareAllDomains
image

File: Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeAdPermissions.ps1 [hunk 1]
```diff
@@ -100,68 +101,74 @@ Function Get-ExchangeAdPermissions {
...
-
-                    if ((($entry.RootOnly) -and ($domainName -eq $rootDomainName)) -or
-                        ($entry.RootOnly -eq $false)) {
-                        Write-Verbose "Trying to find the entry GUID: $($entry.ObjectTypeGuid)"
-                        if ($entry.AdminSdHolder) {
-                            $objectAcl = $adminSdHolderAcl
-                            $objectDN = $adminSdHolderDN
+        $prepareDomainInfo = Get-ExchangeDomainConfigVersion -Domain $domainName
+
+        if ($null -ne $prepareDomainInfo.ObjectVersion) {
```
Comment (dpaulson45) [inline Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeAdPermissions.ps1]: This is not how to test if the domain ever ran `/preparedomain` the MESO container should only be in the root.

Comment (dpaulson45) [reply Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeAdPermissions.ps1]: Or is that just security groups in the root?? I need to look at this again.

Comment (bill-long) [reply Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeAdPermissions.ps1]: image

Comment (lusassl-msft) [reply Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeAdPermissions.ps1]: MESO should be available in every domain where /preparedomain has been run.

Comment (lusassl-msft) [reply Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeAdPermissions.ps1]: > Or is that just security groups in the root?? I need to look at this again.

Yes, the `Microsoft Exchange Security Groups` are available only within the root domain.

File: Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeDomainConfigVersion.ps1 [hunk 2]
```diff
@@ -0,0 +1,47 @@
+﻿# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+Function Get-ExchangeDomainConfigVersion {
```
[the remark this answers is not in the document]
Comment (bill-long) [reply Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeDomainConfigVersion.ps1:4]: This PR is already touching a lot. Let's address that in a new PR.

File: Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeDomainConfigVersion.ps1 [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    if ($forest.Domains.Name.Contains($Domain)) {
+        Write-Verbose "Domain: $Domain is present in forest: $($forest.Name)"
+        $domainObject = $forest.Domains | Where-Object { $_.Name -eq $Domain }
+        $domainDN = $domainObject.GetDirectoryEntry().distinguishedName
+        $adEntry = [ADSI]("LDAP://CN=Microsoft Exchange System Objects," + $domainDN)
+        $sdFinder = New-Object System.DirectoryServices.DirectorySearcher($adEntry)
+        try {
+            $mesoResult = $sdFinder.FindOne()
+        } catch {
+            Write-Verbose "No result was returned"
```
[the remark this answers is not in the document]
Comment (lusassl-msft) [reply Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeDomainConfigVersion.ps1:27]: Done.

File: Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeAdPermissions.ps1 (hunk 1 above)
Comment (lusassl-msft) [reply Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeAdPermissions.ps1]: As discussed with @dpaulson45 we should be good by doing it this way. Resolving the conversation.

**Pull request merged**